addons/hr_holidays/models/hr_leave.py

```diff
@@ -107,45 +107,133 @@ def _default_employee(self):
+                             ('day', '1 Day'),
+                             ('period', 'Period')],
+                            default=lambda req: 'period' if req.leave_type_request_unit == 'hour' else 'half')
+    # Duplicate field because we cannot hide some entries of a selection field
+    request_unit_day = fields.Selection([('day', '1 Day'),
+                                 ('period', 'Period')],
+                                default='day')
+
+    number_of_hours = fields.Float(
+        'Hours Allocation', copy=False, readonly=True, compute='_compute_number_of_hours',
+        help='Number of hours of the leave request according to your working schedule.')
+
     _sql_constraints = [
         ('type_value', "CHECK( (holiday_type='employee' AND employee_id IS NOT NULL) or (holiday_type='category' AND category_id IS NOT NULL) or (holiday_type='department' AND department_id IS NOT NULL) )",
          "The employee, department or employee category of this request is missing. Please make sure that your user login is linked to an employee."),
         ('date_check2', "CHECK ((date_from <= date_to))", "The start date must be anterior to the end date."),
-        ('date_check', "CHECK ( number_of_days_temp >= 0 )", "The number of days must be greater than 0."),
+        ('date_check', "CHECK ( number_of_days_temp >= 0 )", "If you want to change the number of days you should use the 'period' mode"),
     ]
 
+    @api.onchange('request_unit_all', 'request_date_from_period', 'request_date_to_period',
+                  'holiday_status_id', 'request_date_from', 'request_date_to')
+    def _onchange_request_parameters(self):
+        if self.request_date_from:
+            if self.date_from:
+                self.date_from = datetime.combine(fields.Date.from_string(self.request_date_from), fields.Datetime.from_string(self.date_from).time())
+            else:
+                self.date_from = self.request_date_from
+
+        if self.request_date_to:
+            if self.date_to:
+                self.date_to = datetime.combine(fields.Date.from_string(self.request_date_to), fields.Datetime.from_string(self.date_to).time())
+            else:
+                self.date_to = self.request_date_to
+
+        if not self.request_date_from or not self.request_date_to:
+            self.number_of_days_temp = 0
+            return
+
+        date_from = False
+        date_to = False
+
+        domain = [('calendar_id', '=', self.employee_id.resource_calendar_id.id)]
+        attendances = self.env['resource.calendar.attendance'].search(domain, order='dayofweek, day_period DESC')
+
+        first_day = fields.Date.from_string(self.date_from)
+        last_day = fields.Date.from_string(self.date_to)
+
+        if self.request_unit_all in ['day', 'half']:
+            last_day = first_day
+
+        attendance_from = False
+        attendance_to = False
+
+        for weekday in range(first_day.weekday(), 7): # from first_day to next sunday
```

Comment:
I remember we talked about a "next week" or "previous week" attendance fetch. I suppose this code comes from that issue. Couldn't we have something simpler like
```
# find first attendance coming after first_day
attendance_from = next(att for att in attendances if att.dayofweek >= weekday, attendances[0])
```

Reply:
We could but I'm not sure it would be more readable as in order to pass the weekday to the generator expression we have to make a closure

eg:
    attendance_from = next((lambda weekday: (att for att in attendances if int(att.dayofweek) >= weekday))(first_day.weekday()), attendances[0])